### ISSUE TITLE
Glitches in Request Worker.

### ISSUE THREAD

**Issue by dhaval-parekh** [rtCamp/px-wp-insights#190]
1. Request workers have failed the job if "site_title" is empty in request data. Ideally, It needs to allow empty "site_title".

2. During the request job process. the below error is occurring.
```
2021-12-20 13:04 +00:00: Error: data.plugins is not iterable
2021-12-20 13:04 +00:00:     at Queue._onMessage (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/bee-queue/lib/queue.js:138:22)
2021-12-20 13:04 +00:00:     at RedisClient.emit (events.js:315:20)
2021-12-20 13:04 +00:00:     at RedisClient.EventEmitter.emit (domain.js:467:12)
2021-12-20 13:04 +00:00:     at return_pub_sub (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/redis/index.js:786:18)
2021-12-20 13:04 +00:00:     at RedisClient.return_reply (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/redis/index.js:833:9)
2021-12-20 13:04 +00:00:     at JavascriptRedisParser.returnReply (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/redis/index.js:192:18)
```

3. When the user is on the paginated page in the job listing page and clicks on the "Retry" job button, It doesn't work since the POST request is failing because of an invalid request URL.

**Comment by dhaval-parekh** [comment]
4. Request job failing because of foreign key ( `url_error_relationships_error_source_slug_foreign` ) constrain. 

```
2021-12-21 08:08 +00:00: info:  Site: valed.press | Job ID: ampwp-b46a8637-6d90-5f7f-a094-aa381474669d started.
2021-12-21 08:08 +00:00: debug: Site: valed.press reported progress: 15%
2021-12-21 08:08 +00:00: debug: Site: valed.press reported progress: 30%
2021-12-21 08:09 +00:00: debug: Site: valed.press reported progress: 45%
2021-12-21 08:09 +00:00: debug: Site: valed.press reported progress: 50%
2021-12-21 08:09 +00:00: debug: Site: valed.press reported progress: 70%
2021-12-21 08:09 +00:00: crit: Site: valed.press | Job ID: ampwp-b46a8637-6d90-5f7f-a094-aa381474669d failed. 
2021-12-21 08:09 +00:00:       Error: ER_NO_REFERENCED_ROW_2: Cannot add or update a child row: a foreign key constraint fails (`amp_compatibility_server`.`url_error_relationships`, CONSTRAINT `url_error_relationships_error_source_slug_foreign` FOREIGN KEY (`error_source_slug`) REFERENCES `error_sources` (`error_source)
2021-12-21 08:09 +00:00:     at Query.Sequence._packetToError (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/mysql/lib/protocol/sequences/Sequence.js:47:14)
2021-12-21 08:09 +00:00:     at Query.ErrorPacket (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/mysql/lib/protocol/sequences/Query.js:79:18)
2021-12-21 08:09 +00:00:     at Protocol._parsePacket (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/mysql/lib/protocol/Protocol.js:291:23)
2021-12-21 08:09 +00:00:     at Parser._parsePacket (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/mysql/lib/protocol/Parser.js:433:10)
2021-12-21 08:09 +00:00:     at Parser.write (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/mysql/lib/protocol/Parser.js:43:10)
2021-12-21 08:09 +00:00:     at Protocol.write (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/mysql/lib/protocol/Protocol.js:38:16)
2021-12-21 08:09 +00:00:     at TLSSocket.ondata (internal/streams/readable.js:719:22)
2021-12-21 08:09 +00:00:     at TLSSocket.emit (events.js:315:20)
2021-12-21 08:09 +00:00:     at TLSSocket.EventEmitter.emit (domain.js:467:12)
2021-12-21 08:09 +00:00:     at addChunk (internal/streams/readable.js:309:12)
2021-12-21 08:09 +00:00:     at readableAddChunk (internal/streams/readable.js:284:9)
2021-12-21 08:09 +00:00:     at TLSSocket.Readable.push (internal/streams/readable.js:223:10)
2021-12-21 08:09 +00:00:     at TLSWrap.onStreamRead (internal/stream_base_commons.js:188:23)
2021-12-21 08:09 +00:00:     --------------------
2021-12-21 08:09 +00:00:     at Protocol._enqueue (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/mysql/lib/protocol/Protocol.js:144:48)
2021-12-21 08:09 +00:00:     at Connection.query (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/mysql/lib/Connection.js:198:25)
2021-12-21 08:09 +00:00:     at /opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/knex/lib/dialects/mysql/index.js:134:18
2021-12-21 08:09 +00:00:     at new Promise (<anonymous>)
2021-12-21 08:09 +00:00:     at Client_MySQL._query (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/knex/lib/dialects/mysql/index.js:128:12)
2021-12-21 08:09 +00:00:     at Client_MySQL.query (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/knex/lib/client.js:168:17)
2021-12-21 08:09 +00:00:     at Runner.query (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/knex/lib/runner.js:151:36)
2021-12-21 08:09 +00:00:     at /opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/knex/lib/runner.js:40:23
2021-12-21 08:09 +00:00:     at /opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/knex/lib/runner.js:277:24
2021-12-21 08:09 +00:00:     at runMicrotasks (<anonymous>)
2021-12-21 08:09 +00:00:     at processTicksAndRejections (internal/process/task_queues.js:93:5)
2021-12-21 08:09 +00:00:     at async Proxy._insert (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/node_modules/@adonisjs/lucid/src/Lucid/Model/index.js:612:20)
2021-12-21 08:09 +00:00:     at async Function.save (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/app/Models/Base.js:197:12)
2021-12-21 08:09 +00:00:     at async Function.saveValidatedUrls (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/app/Controllers/Queue/RequestController.js:615:46)
2021-12-21 08:09 +00:00:     at async Function.processJob (/opt/workspace/px-wp-insights/repo/px-wp-insights-server/app/Controllers/Queue/RequestController.js:272:29) {
2021-12-21 08:09 +00:00:   code: 'ER_NO_REFERENCED_ROW_2',
2021-12-21 08:09 +00:00:   errno: 1452,
2021-12-21 08:09 +00:00:   sqlMessage: 'Cannot add or update a child row: a foreign key constraint fails (`amp_compatibility_server`.`url_error_relationships`, CONSTRAINT `url_error_relationships_error_source_slug_foreign` FOREIGN KEY (`error_source_slug`) REFERENCES `error_sources` (`error_source)',
2021-12-21 08:09 +00:00:   sqlState: '23000',
2021-12-21 08:09 +00:00:   index: 0,
2021-12-21 08:09 +00:00:   sql: "insert into `url_error_relationships` (`error_slug`, `error_source_slug`, `hash`, `page_url`, `site_url`) values ('718ba72a751cd393d5c76675474694d3534957bc315f53433779cd155793b837', '579872f482004044c09a7c943f7d3e9f2ac333e16407bc4ff3e4062ed5de198a', '423c0ad16483addedaea434efe0d25dbd0c6e5e3cc472c6c5eb2f83344806db7', 'valed.press/overseas-education/valentaines-day-wishes-and-messages/amp', 'valed.press')"
2021-12-21 08:09 +00:00: }
2021-12-21 08:09 +00:00: info: Retrying: Site: valed.press | Job ID: ampwp-b46a8637-6d90-5f7f-a094-aa381474669d
```

Ideally, Error source information already be in the error source table before inserting

5. On the request job listing page, the "Copy Data" button copy minified version of data instead it should copy the full version of data.

**Comment by dhaval-parekh** [comment]
> 4. Request job failing because of foreign key ( url_error_relationships_error_source_slug_foreign ) constrain.

Jobs are failing because there are some error sources that don't have a `type` field. For example.

```json
   "183": {
      "error_slug": "a495346904db5354cfc9ef88fe21daa7f098cbea7c7aebb6c0331242446f0201",
      "error_source_slug": "f7115bf7f1084fbb9283d09783cc5dca23de15b5fa20375fb95facf7566dcaaa",
      "filter": "1",
      "hook": "the_content",
      "post_id": "18618",
      "post_type": "page",
      "sources": [
        {
          "error_source_slug": "838b521f1310fe679b126eb33b79d2e093d2303e334cf2fc96980f5c3666781f",
          "file": "includes/frontend.php",
          "function": "Elementor\\Frontend::apply_builder_in_content",
          "line": "995",
          "name": "elementor",
          "type": "plugin",
          "version": "3.5.1"
        },
        {
          "error_source_slug": "53d8f6e99dae0cd05eec7ec997c59d7bd8dc7f8a6c2375c87fe03890c51b4286",
          "file": "includes/class-frontend-seo-score.php",
          "function": "RankMath\\Frontend_SEO_Score::insert_score",
          "line": "67",
          "name": "seo-by-rank-math",
          "type": "plugin",
          "version": "1.0.78"
        },
        {
          "error_source_slug": "44221646ddd5d4893b8243879ddd6aac5d143d7462ead7da8555c6ec392b64b7",
          "file": "includes/Core/Util/Method_Proxy_Trait.php",
          "function": "Google\\Site_Kit\\Core\\Util\\{closure}",
          "line": "24",
          "name": "google-site-kit",
          "type": "plugin",
          "version": "1.48.0"
        },
        {
          "error_source_slug": "83bafe0b0bfa4df67fdcde6ba83842a63fec2c37fe4805324fdae6ca633f5c9a",
          "file": "includes/modules/schema/class-snippet-shortcode.php",
          "function": "RankMath\\Schema\\Snippet_Shortcode::output_schema_in_content",
          "line": "362",
          "name": "seo-by-rank-math",
          "type": "plugin",
          "version": "1.0.78"
        },
        {
          "error_source_slug": "38984d4cd3478a8b9d3087f5673506c871c2ad93510cfbb06d488449dbb82529",
          "file": "includes/frontend/class-link-attributes.php",
          "function": "RankMath\\Frontend\\Link_Attributes::add_link_attributes",
          "line": "60",
          "name": "seo-by-rank-math",
          "type": "plugin",
          "version": "1.0.78"
        }
      ],
      "type": "",
      "version": "0.0"
    },
```

We don't allow the type to be empty since from the type we generate a foreign key value for `error_source.extension_version_slug` to `extension_version.extension_version_slug`. If we allow `type` to empty then it won't be able to generate value for `extension_version_slug` so instead of allowing type to empty we take a value of `version`, `type`, and `name` from the child source.

**Comment by dhaval-parekh** [comment]
6. Request jobs are failing because of foreign key `error_sources_extension_version_slug_foreign` constrain. [reference](https://storage.cloud.google.com/amp_comp_db/logs/request-worker/ampwp-ea4fbf32-29cd-522b-ad01-dc78a9e9b933.log)

Ideally, the extension version should already be in the `extension_versions` table.

**Comment by dhaval-parekh** [comment]
> 6. Request jobs are failing because of foreign key error_sources_extension_version_slug_foreign constrain.

This has been addressed in the commit [4158ed8](https://github.com/rtCamp/px-wp-insights/commit/4158ed808ab13af5a41ad15c9da44f9154f543ab)
